core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -241,6 +246,58 @@ static MetaData updateMetaDataWithRoutingTable(MetaData oldMetaData, RoutingTabl
     }
 
     /**
+     * Possibly removes delay markers from shards and reroutes if markers have been deleted.
+     * <p>
+     * If the same instance of the routing table is returned, then no change has been made.</p>
+     */
+    public RoutingAllocation.Result applyDelayedShards(ClusterState clusterState) {
+        return applyDelayedShards(clusterState, true);
+    }
+
+    public RoutingAllocation.Result applyDelayedShards(ClusterState clusterState, boolean withReroute) {
+        RoutingNodes routingNodes = getMutableRoutingNodes(clusterState);
+        RoutingAllocation allocation = new RoutingAllocation(allocationDeciders, routingNodes, clusterState.nodes(), clusterInfoService.getClusterInfo(), currentNanoTime());
+        boolean changed = removeDelayMarkers(allocation);
+        if (changed == false) {
+            return new RoutingAllocation.Result(false, clusterState.routingTable(), clusterState.metaData());
+        }
+        if (withReroute) {
+            reroute(allocation);
+        }
+        RoutingAllocation.Result result = buildChangedResult(allocation);
+        logClusterHealthStateChange(
```

Comment:
we always log this after building the change result, maybe fold it into the buildChangedResult method and call it buildResultAndLogHealthChange?